### PR TITLE
Fix sub-stride OBB loss divergence via probiou floor

### LINKED ISSUE #24920: Train obb model difficult to convergr
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

When training an OBB model on my own dataset with version 8.4.47, I found it difficult to converge. The first few epochs were normal, but later the loss became NaN. However, when I trained with version 8.4.12 previously, it worked fine. So, I compared the training performance of versions 8.4.12 and 8.4.60 (the latest version at the time of comparison) on DOTA v1.0 and found that the training results of version 8.4.60 indeed seemed problematic. Here are my experimental steps:  
1. First, I used the following code to crop the DOTA v1.0 data. (Correspondingly, I modified the path in DOTAv1.yaml to DOTAv1.0-split.)  
```
from ultralytics.data.split_dota import split_test, split_trainval

# Split train and val set, with labels.
split_trainval(
    data_root="path/to/DOTAv1.0/",
    save_dir="path/to/DOTAv1.0-split/",
    rates=[0.5, 1.0, 1.5],  # multiscale
    gap=500,
)
# Split test set, without labels.
split_test(
    data_root="path/to/DOTAv1.0/",
    save_dir="path/to/DOTAv1.0-split/",
    rates=[0.5, 1.0, 1.5],  # multiscale
    gap=500,
```

3. I trained using the same code.  
4. Below are the training results for the two versions. For version 8.4.12, the best model achieved mAP50: 0.745 and mAP50-95: 0.597, which appears normal. However, for version 8.4.60, the best model only achieved mAP50: 0.548 and mAP50-95: 0.405, and the last model only reached 0.0943 and 0.080.

model  | mAP50 | mAP50-95
-- | -- | --
8.4.12 best | 0.745 | 0.597
 8.4.60 best | 0.548 | 0.405
 8.4.60 last | 0.0943 |  0.080

V8.4.12
Image
V8.4.60
Image

### Environment

Ultralytics 8.4.60 🚀 Python-3.10.14 torch-2.5.1+cu118 CUDA:0 (NVIDIA A40, 45434MiB)
Setup complete ✅ (255 CPUs, 503.2 GB RAM, 0.0/10.0 GB disk)

OS                     Linux-5.4.132-2.pclimit.el7.x86_64-x86_64-with-glibc2.35
Environment            Docker
Python                 3.10.14
Install                git
Path                   /netds2/groupdata/medcvocralgorithm/10_Private/wang_fei/00_work/ultralytics-v8.4.60/ultralytics
RAM                    503.22 GB
Disk                   0.0/10.0 GB
CPU                    AMD EPYC 7713 64-Core Processor
CPU count              255
GPU                    NVIDIA A40, 45434MiB
GPU count              1
CUDA                   11.8

numpy                  ✅ 1.26.4>=1.23.0
matplotlib             ✅ 3.9.2>=3.3.0
opencv-python          ✅ 4.10.0.84>=4.6.0
pillow                 ✅ 11.0.0>=7.1.2
pyyaml                 ✅ 6.0.1>=5.3.1
requests               ✅ 2.32.3>=2.23.0
scipy                  ✅ 1.14.1>=1.4.1
torch                  ✅ 2.5.1+cu118>=1.8.0
torch                  ✅ 2.5.1+cu118!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.20.1+cu118>=0.9.0
psutil                 ✅ 5.9.8>=5.8.0
polars                 ✅ 1.35.2>=0.20.0
ultralytics-thop       ✅ 2.0.19>=2.0.18

### Minimal Reproducible Example
```
import os
# os.environ['TORCH_DISTRIBUTED_DEBUG'] = 'DETAIL'
os.environ['NCCL_BLOCKING_WAIT'] = '0'  
os.environ['YOLO_CONFIG_DIR'] = os.path.dirname(os.path.abspath(__file__))
os.environ['YOLO_OFFLINE'] = 'true' 

os.system('yolo checks')

import argparse
import ultralytics
from ultralytics import YOLO
from datetime import datetime
from pathlib import Path
import shutil

def main():
# -------------------------------------------------------基本配置---------------------------------------------------------------
    project_root = Path(__file__).resolve().parent.parent
    # 路径配置
    model_cfg = project_root / f"ultralytics/cfg/models/26/yolo26-obb.yaml"
    data_yaml = project_root / f"ultralytics/cfg/datasets/DOTAv1.yaml"
 #  ##############################################################################################################################  
    # 保存路径
    project_name = project_root / f"runs/{ultralytics.__version__}"
    project_name.mkdir(parents=True, exist_ok=True)
    print(f"✅✅✅✅✅✅✅✅✅✅\033[33;41m {project_name} \033[0m✅✅✅✅✅✅✅✅✅✅")
    
# -------------------------------------------------------训练参数配置---------------------------------------------------------------
    cfg=dict(
        project=str(project_name),
        data=str(data_yaml),
        epochs=300,
        batch=32,
        imgsz=1024,
        
        # 其他可选配置
        workers=16,
        scale=1.0,
        mosaic=0.0,
        device="0",
        multi_scale=0.0,
        save_period=1,
        resume=False,
        amp=True,
        patience=0,
        seed=1,
        iou= 0.45,
        freeze=None,
    )
    #  ################################################################################################################################     
    # 训练 
    model = YOLO(str(model_cfg))
    model.train(**cfg)
    print(f"✅✅✅✅✅✅✅✅✅✅\033[33;41m {project_name} \033[0m✅✅✅✅✅✅✅✅✅✅")

if __name__ == "__main__":
    main()
```
### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by lmycross**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Improves rotated bounding box training stability for OBB models by adding gradient-safe handling for very small boxes 📦✨

### 📊 Key Changes
- Added a `floor` parameter to rotated box covariance calculations in `ultralytics/utils/metrics.py`.
- Updated `probiou()` to pass this floor value into covariance matrix generation.
- Set a default `floor = 0.01` in `RotatedBboxLoss` and applied it during OBB IoU loss calculation.
- Removed previous width/height masking logic that forcibly adjusted small ground-truth boxes during OBB loss preprocessing.

### 🎯 Purpose & Impact
- Helps prevent unstable gradients when training with very small, sub-stride oriented bounding boxes 🛡️
- Makes probabilistic IoU for rotated boxes more numerically stable without modifying ground-truth box dimensions.
- Improves OBB training reliability, especially for datasets containing tiny rotated objects.
- Keeps the change focused on loss/metric computation, reducing unintended side effects from altering annotation-derived box sizes.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @lmycross, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message is here to help with the initial review, and an Ultralytics engineer will assist further.

Your PR appears focused on improving OBB loss stability for sub-stride boxes via a `probiou` floor—thanks for the targeted fix! 🔧 To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

PR looks clean. The covariance floor is threaded through `probiou()` without breaking existing default behavior, and the OBB loss path now applies it explicitly while preserving original target dimensions.

**Comment by UltralyticsAssistant** [comment]
Merged — thank you @lmycross! 🚀

“Great things are done by a series of small things brought together.” — Vincent van Gogh

This PR is a great example of that: a focused numerical-stability improvement that can make a meaningful difference for OBB training, especially with tiny rotated objects. By adding gradient-safe handling in rotated box covariance and `probiou()` without altering ground-truth dimensions, this helps improve training reliability while keeping the loss behavior cleaner and more principled.

Appreciate the thoughtful contribution to making Ultralytics OBB models more robust! 🙌
